repos/system_upgrade/el8toel9/actors/checkvdo/libraries/checkvdo.py

```diff
@@ -0,0 +1,180 @@
+    return [_canonicalize_device_path(x) for x in results.splitlines()]
+
+
+#
+# Methods used in discovering VDO devices for which migration to lvm-based
+# management was attempted but not completed.
+#
+def _get_migration_failed_blkid_results():
+    """ Collect all devices blkid identifies. """
+    # VDO devices that have been created by or migrated to lvm-based management
+    # are not identified by blkid.
+    command = ['blkid', '--output', 'device']
+    return _run_cmd(command)['stdout']
+
+
+def _get_migration_failed_canonical_blkid_results():
+    """ Collect canonical devices blkid identifies. """
+    results = _get_migration_failed_blkid_results()
+    # Canonicalize the device paths.
+    return [_canonicalize_device_path(x) for x in results.splitlines()]
+
+
+def _get_migration_failed_lsblk_results():
+    """ Collect all devices known to the system. """
+    # lsblk does not identify lvm-managed vdos as vdos.
+    # Any such devices will be eliminated by the check for being a
+    # post-migration vdo as the check will identify them as pre-migration vdos.
+    command = ['lsblk', '--noheadings', '--output', 'NAME,TYPE', '--path', '--list']
+    return _run_cmd(command)['stdout']
+
+
+def _get_migration_failed_lsblk_info():
+    """ Collect canonical devices lsblk identifies. """
+    results = _get_migration_failed_lsblk_results()
+    info = {}
+    for line in results.splitlines():
+        # Use 'kind' for type as 'type' is a python entity.
+        (device, kind) = line.split()
+        device = _canonicalize_device_path(device)
+        info[device] = { "type" : kind }
+    return info
+
+
+def _is_post_migration_vdo_device(device):
+    """ Identify if the specified device is a post-migration vdo device. """
+    # A post-migration vdo device is a vdo device which has completed its
+    # preparation for being managed by lvm but for which lvm has failed (for
+    # whatever reason; e.g., system crash) to take up that responsibility.
+    #
+    # N.B. lvm-managed vdos will make it to the point of being checked here.
+    #      This is due to the fact that they are not recognized by either
+    #      blkid or lsblk as vdo devices and are thus not eliminated using the
+    #      results of those.
+    #      They are, though, recognized by the conversion utility as
+    #      pre-migration vdo devices and are thus eliminated by this check.
+    conversion_utility = os.path.join(os.getenv('VDO_CONVERSION_UTILITY_DIR',
+                                                '/usr/libexec'),
+                                      'vdoprepareforlvm')
+    command = [conversion_utility, '--check', device]
+    result = _run_cmd(command, checked = False)
+    exit_code = result['exit_code']
+    # 255: Not a vdo device
+    #   0: A post-migration vdo device
+    #   1: A pre-migration vdo device
+    if (exit_code != 255) and (exit_code != 0) and (exit_code != 1):
+        raise exceptions.StopActorExecutionError(
```

Comment:
Again, this is quite harsh please make this an inhibitor instead

Reply:
@vinzenz 
Addressed in commit 465a92f via generation of inhibitory report.